### PR TITLE
Obvious maneuvers

### PR DESCRIPTION
**Pull request by dgearhart**
# Issue

Identify and collapse obvious maneuvers

Fixes  https://github.com/valhalla/valhalla/issues/1744
Fixes https://github.com/valhalla/valhalla/issues/91
Fixes https://github.com/valhalla/valhalla/issues/82

## Stats for 1k user routes

- Improved 67% of the route narratives
- Reduced maneuvers by 12%

## Example#1 
``` diff
< BEFORE
< 2: Turn left onto US 422/West Chocolate Avenue.
< 3: Continue on US 322 West.
< 4: Take the exit toward Middletown/Hummelstown.
< 5: Turn left onto Quarry Road.
< 6: Continue on Waltonville Road.
< 7: You have arrived at your destination.
> AFTER
> 2: Turn left onto US 422/West Chocolate Avenue.
> 3: Take the exit toward Middletown/Hummelstown.
> 4: Turn left onto Quarry Road.
> 5: You have arrived at your destination.
```
image

## Example#2
``` diff
< BEFORE
< 2: Take exit 31C on the left onto White Marsh Boulevard toward White Marsh.
< 3: Continue on MD 43/White Marsh Boulevard.
< 4: Turn right onto Perry Hills Court.
> AFTER
> 2: Take exit 31C on the left onto MD 43 East/White Marsh Boulevard toward White Marsh.
> 3: Turn right onto Perry Hills Court.
```

image

## Example#3
``` diff
< BEFORE
< 2: Turn left onto Britzer Damm.
< 3: Continue on Britzer Brücke.
< 4: Continue on Britzer Damm.
< 5: Turn left onto Franz-Körner-Straße.
> AFTER
> 2: Turn left onto Britzer Damm.
> 3: Turn left onto Franz-Körner-Straße.
```

image

## Tasklist

 - [x] Add RAD tests
 - [x] Update gurka tests
 - [x] Add pinpoint tests
 - [x] Add gurka tests
 - [x] Add #fixes with the issue number that this PR addresses
 - [x] Generally use squash merge to rebase and clean comments before merging
 - [x] Update the [changelog](CHANGELOG.md)

### CODE REVIEW

File: src/odin/enhancedtrippath.cc [hunk 1]
```diff
@@ -1297,6 +1300,24 @@ bool EnhancedTripLeg_Node::HasIntersectingEdgeCurrNameConsistency() const {
   return false;
 }
 
+bool EnhancedTripLeg_Node::HasNonBackwardTraversableSameNameIntersectingEdge(
+    uint32_t from_heading,
```
Comment (mandeepsandhu) [inline src/odin/enhancedtrippath.cc:1304]: Indentation looks off.

File: src/odin/enhancedtrippath.cc [hunk 2]
```diff
@@ -1297,6 +1300,24 @@ bool EnhancedTripLeg_Node::HasIntersectingEdgeCurrNameConsistency() const {
   return false;
 }
 
+bool EnhancedTripLeg_Node::HasNonBackwardTraversableSameNameIntersectingEdge(
```
Comment (mandeepsandhu) [inline src/odin/enhancedtrippath.cc:1303]: Can you add some comments on what this function does?

Comment (dgearhart) [reply src/odin/enhancedtrippath.cc:1303]: @mandeepsandhu i added comments

File: src/odin/maneuversbuilder.cc [hunk 3]
```diff
@@ -2421,8 +2519,14 @@ void ManeuversBuilder::ProcessRoundaboutNames(std::list<Maneuver>& maneuvers) {
       // Process roundabout exit names and signs
       if (next_man->type() == DirectionsLeg_Maneuver_Type_kRoundaboutExit) {
         if (next_man->HasBeginStreetNames()) {
-          curr_man->set_roundabout_exit_begin_street_names(next_man->begin_street_names().clone());
-          curr_man->set_roundabout_exit_street_names(next_man->street_names().clone());
+          if (next_man->contains_obvious_maneuver()) {
```
Comment (danpaz) [inline src/odin/maneuversbuilder.cc:2522]: Any concerns that this will cause issues when we merge it with #2437 ?

Comment (dgearhart) [reply src/odin/maneuversbuilder.cc:2522]: No issues

File: test/gurka/test_instructions_turn_toward.cc [hunk 4]
```diff
@@ -252,18 +252,19 @@ TEST_F(InstructionsTurnToward, LeftUturnTowardUsingInternalEdges) {
...
-                                                DirectionsLeg_Maneuver_Type_kDestination});
-  int maneuver_index = 1;
-
-  // Verify the turn left toward instructions
-  gurka::assert::raw::expect_instructions_at_maneuver_index(
-      result, maneuver_index, "Continue toward B1/C1/Little Italy.", "Continue toward B1.",
-      "Continue toward B1, Little Italy. Then You will arrive at your destination.",
-      "Continue for 200 meters.");
-}
+// TODO: expand map for obvious maneuver
```
Comment (danpaz) [inline test/gurka/test_instructions_turn_toward.cc:255]: Is this still a TODO @dgearhart ?

File: src/odin/maneuversbuilder.cc [hunk 5]
```diff
@@ -2369,6 +2397,76 @@ bool ManeuversBuilder::AreRampManeuversCombinable(std::list<Maneuver>::iterator
   return false;
 }
 
+bool ManeuversBuilder::IsNextManeuverObvious(std::list<Maneuver>& maneuvers,
```
Comment (mandeepsandhu) [inline src/odin/maneuversbuilder.cc:2400]: I think all the args can be `const` refs? Or are we modifying them somewhere within (I couldn't find any place where we were, but I might've missed something)?

File: src/odin/maneuversbuilder.cc [hunk 6]
```diff
@@ -2369,6 +2397,76 @@ bool ManeuversBuilder::AreRampManeuversCombinable(std::list<Maneuver>::iterator
...
+            return true;
+          }
+        }
+      }
+      return false;
+    }
+
+    // Return true if a short continue maneuver
+    // and the following maneuver is not a continue
+    if (next_man->length(Options_Units_kilometers) < kShortContinueThreshold) {
```
Comment (mandeepsandhu) [inline src/odin/maneuversbuilder.cc:2448]: Could this function be called with `next_man` pointing to the last/arrival maneuver? If so, should we return `false` if `next_next_man == maneuvers.end()` ?

File: test/gurka/test_instructions_obvious_maneuver.cc [hunk 7]
```diff
@@ -0,0 +1,128 @@
+#include "gurka.h"
```
Comment (mandeepsandhu) [inline test/gurka/test_instructions_obvious_maneuver.cc:1]: Thank you for these tests! Its really easy to visualize how the continue collapsing is working! 🙏

File: test/gurka/test_instructions_turn_toward.cc (hunk 4 above)
Comment (dgearhart) [reply test/gurka/test_instructions_turn_toward.cc:255]: yes

File: src/odin/maneuversbuilder.cc (hunk 6 above)
Comment (dgearhart) [reply src/odin/maneuversbuilder.cc:2448]: `next_man` must be a `DirectionsLeg_Maneuver_Type_kContinue`

File: src/odin/maneuversbuilder.cc (hunk 5 above)
Comment (dgearhart) [reply src/odin/maneuversbuilder.cc]: made params const

File: src/odin/maneuversbuilder.cc (hunk 3 above)
Comment (dgearhart) [reply src/odin/maneuversbuilder.cc:2522]: @danpaz I tested - this looks good

**Pull request merged**